derive/src/utils.rs

```diff
@@ -61,63 +66,81 @@ pub fn produce_signature<T: quote::ToTokens>(
+		syn::Expr::Lit(syn::ExprLit{lit: syn::Lit::Int(lit_int), ..}) => {
+			target.push_str(&format!("{}", lit_int.value()))
+		}
 		_ => panic!("Cannot use something other than integer literal in this constant expression"),
 	}
 }
 
-pub fn push_canonical(target: &mut String, ty: &syn::Ty) {
-	match *ty {
-		syn::Ty::Path(None, ref path) => {
-			let last_path = path.segments.last().unwrap();
-			match last_path.ident.to_string().as_ref() {
-				"u32" => target.push_str("uint32"),
-				"i32" => target.push_str("int32"),
-				"u64" => target.push_str("uint64"),
-				"i64" => target.push_str("int64"),
-				"U256" => target.push_str("uint256"),
-				"H256" => target.push_str("uint256"),
-				"Address" => target.push_str("address"),
-				"Vec" => {
-					match last_path.parameters {
-						syn::PathParameters::AngleBracketed(ref param_data) => {
-							let vec_arg = param_data.types.last().unwrap();
-							if let syn::Ty::Path(None, ref nested_path) = *vec_arg {
-								if "u8" == nested_path.segments.last().unwrap().ident.to_string() {
-									target.push_str("bytes");
-									return;
-								}
-							}
-							push_canonical(target, vec_arg);
-							target.push_str("[]")
-						},
-						_ => panic!("Unsupported vec arguments"),
-					}
-				},
-				"String" => target.push_str("string"),
-				"bool" => target.push_str("bool"),
-				ref val @ _ => panic!("Unable to handle param of type {}: not supported by abi", val)
+fn push_canonical_vec(target: &mut String, args: &syn::PathArguments) {
+	match args {
+		syn::PathArguments::AngleBracketed(gen_args) => {
+			let last_arg = gen_args.args.last().unwrap();
+			let last_type = last_arg.value();
+			if let syn::GenericArgument::Type(syn::Type::Path(type_path)) = last_type {
+				return if type_path.qself.is_none()
+					&& type_path.path.segments.last().unwrap().value().ident == "u8"
+				{
+					target.push_str("bytes");
+				}
+				else {
+					push_canonical_path(target, type_path);
+					target.push_str("[]");
+				}
 			}
+			panic!("Unsupported generic arguments")
 		},
-		syn::Ty::Array(ref nested_ty, ref const_expr) => {
-			// special cases for bytesXXX
-			if let syn::Ty::Path(None, ref nested_path) = **nested_ty {
-				if "u8" == nested_path.segments.last().unwrap().ident.to_string() {
+		_ => panic!("Unsupported vec arguments"),
+	}
+}
+
+fn push_canonical_primitive(target: &mut String, seg: &syn::PathSegment) {
+	match seg.ident.to_string().as_str() {
+		"u32"     => target.push_str("uint32"),
+		"i32"     => target.push_str("int32"),
+		"u64"     => target.push_str("uint64"),
+		"i64"     => target.push_str("int64"),
+		"U256"    => target.push_str("uint256"),
+		"H256"    => target.push_str("uint256"),
+		"Address" => target.push_str("address"),
+		"String"  => target.push_str("string"),
+		"bool"    => target.push_str("bool"),
+		"Vec"     => push_canonical_vec(target, &seg.arguments),
+		val       => panic!("[e1] Unable to handle param of type {}: not supported by abi", val)
```

Comment:
Vertical alignment isn't something that we have in our house style. Certainly not mixing spaces and tabs : )

Reply:
I can understand that. I will revert the extra vertical alignment.